Node flake in interop test

-- deleted bad log

Final output:
```
  1 failing
  1) Interop tests should pass server_streaming:
     Uncaught AssertionError: 3 === 4
      at ClientReadableStream.<anonymous> (/home/travis/build/grpc/grpc/src/node/interop/interop_client.js:159:12)
      at ClientReadableStream.emit (events.js:95:17)
      at /home/travis/build/grpc/grpc/src/node/src/client.js:386:16
```

Full log here:
https://s3.amazonaws.com/archive.travis-ci.org/jobs/63581683/log.txt

My suspicion is that call.c is discarding the final message errantly.

Is there a way to run just the server_streaming interop test?

Complete log of a failure:
```
+ CONFIG=opt
++ dirname tools/run_tests/run_node.sh
+ cd tools/run_tests/../..
++ pwd
+ root=/usr/local/google/home/ctiller/grpc
+ export LD_LIBRARY_PATH=/usr/local/google/home/ctiller/grpc/libs/opt
+ LD_LIBRARY_PATH=/usr/local/google/home/ctiller/grpc/libs/opt
+ /usr/local/google/home/ctiller/grpc/src/node/node_modules/mocha/bin/mocha --timeout 8000 /usr/local/google/home/ctiller/grpc/src/node/test --grep server_streaming


  Interop tests
Server starting
Serving /grpc.testing.TestService/EmptyCall
Serving /grpc.testing.TestService/UnaryCall
Serving /grpc.testing.TestService/StreamingOutputCall
Serving /grpc.testing.TestService/StreamingInputCall
Serving /grpc.testing.TestService/FullDuplexCall
Serving /grpc.testing.TestService/HalfDuplexCall
I0522 09:09:44.986836458   16441 call.c:1199] grpc_call_start_batch(call=0x20582c0, ops=0x2053c50, nops=4, tag=0x2053f30)
I0522 09:09:44.986926648   16441 call.c:1199] ops[0]: SEND_INITIAL_METADATA
I0522 09:09:44.986937975   16441 call.c:1199] ops[1]: SEND_MESSAGE ptr=0x2053d60
I0522 09:09:44.986944634   16441 call.c:1199] ops[2]: SEND_CLOSE_FROM_CLIENT
I0522 09:09:44.986951153   16441 call.c:1199] ops[3]: RECV_INITIAL_METADATA ptr=0x2053f08
I0522 09:09:44.987040553   16441 call.c:1199] grpc_call_start_batch(call=0x20582c0, ops=0x2053c50, nops=1, tag=0x2054180)
I0522 09:09:44.987053396   16441 call.c:1199] ops[0]: RECV_STATUS_ON_CLIENT metadata=0x2053ca8 status=0x2053cc0 details=0x2053cc8
I0522 09:09:44.987711858   16441 call.c:1199] grpc_call_start_batch(call=0x20582c0, ops=0x2053c50, nops=1, tag=0x20540b0)
I0522 09:09:44.987729726   16441 call.c:1199] ops[0]: RECV_MESSAGE ptr=0x2053fb8
I0522 09:09:45.784953053   17075 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x204fcb0 OK
I0522 09:09:45.785643755   16441 call.c:1199] grpc_call_start_batch(call=0x7ff53c000930, ops=0x20573f0, nops=1, tag=0x2059300)
I0522 09:09:45.785660732   16441 call.c:1199] ops[0]: RECV_CLOSE_ON_SERVER cancelled=0x20592a8
I0522 09:09:45.785681268   16441 call.c:1199] grpc_call_start_batch(call=0x7ff53c000930, ops=0x20594e0, nops=2, tag=0x20573f0)
I0522 09:09:45.785688496   16441 call.c:1199] ops[0]: SEND_INITIAL_METADATA
I0522 09:09:45.785693462   16441 call.c:1199] ops[1]: RECV_MESSAGE ptr=0x20595c8
I0522 09:09:45.785731777   17075 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x20573f0 OK
I0522 09:09:45.785774251   17074 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x2053f30 OK
I0522 09:09:45.787698190   16441 call.c:1199] grpc_call_start_batch(call=0x7ff53c000930, ops=0x204fcb0, nops=1, tag=0x2059980)
I0522 09:09:45.787720972   16441 call.c:1199] ops[0]: SEND_MESSAGE ptr=0x2059810
I0522 09:09:45.787756738   17074 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x2059980 OK
I0522 09:09:45.788648408   17074 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x20540b0 OK
I0522 09:09:45.788777430   16441 call.c:1199] grpc_call_start_batch(call=0x7ff53c000930, ops=0x204fcb0, nops=1, tag=0x20593e0)
I0522 09:09:45.788790514   16441 call.c:1199] ops[0]: SEND_MESSAGE ptr=0x2053e50
I0522 09:09:45.788842434   17075 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x20593e0 OK
I0522 09:09:45.789545604   16441 call.c:1199] grpc_call_start_batch(call=0x20582c0, ops=0x204fcb0, nops=1, tag=0x2059770)
I0522 09:09:45.789558787   16441 call.c:1199] ops[0]: RECV_MESSAGE ptr=0x20595a8
I0522 09:09:45.789592130   17073 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x2059770 OK
I0522 09:09:45.789782950   16441 call.c:1199] grpc_call_start_batch(call=0x7ff53c000930, ops=0x204fcb0, nops=1, tag=0x20597a0)
I0522 09:09:45.789795153   16441 call.c:1199] ops[0]: SEND_MESSAGE ptr=0x2054b70
I0522 09:09:45.789843926   17075 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x20597a0 OK
I0522 09:09:45.790066266   16441 call.c:1199] grpc_call_start_batch(call=0x20582c0, ops=0x20573f0, nops=1, tag=0x2059480)
I0522 09:09:45.790078186   16441 call.c:1199] ops[0]: RECV_MESSAGE ptr=0x2053f48
I0522 09:09:45.790107984   17075 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x2059480 OK
I0522 09:09:45.790283761   16441 call.c:1199] grpc_call_start_batch(call=0x7ff53c000930, ops=0x2059770, nops=1, tag=0x2059370)
I0522 09:09:45.790300585   16441 call.c:1199] ops[0]: SEND_MESSAGE ptr=0x2059bc0
I0522 09:09:45.790326006   17074 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x2059370 OK
I0522 09:09:45.791176705   16441 call.c:1199] grpc_call_start_batch(call=0x20582c0, ops=0x20594b0, nops=1, tag=0x20540b0)
I0522 09:09:45.791190700   16441 call.c:1199] ops[0]: RECV_MESSAGE ptr=0x2054128
I0522 09:09:45.791302415   16441 call.c:1199] grpc_call_start_batch(call=0x7ff53c000930, ops=0x204fcb0, nops=1, tag=0x2059980)
I0522 09:09:45.791313447   16441 call.c:1199] ops[0]: SEND_STATUS_FROM_SERVER status=0 details=OK
I0522 09:09:45.791348059   17074 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x2059980 OK
I0522 09:09:45.791374491   17072 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x2059300 OK
I0522 09:09:45.791953174   17074 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x2054180 OK
I0522 09:09:45.791954262   17075 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x20540b0 OK
    1) should pass server_streaming
I0522 09:09:45.793317403   16441 call.c:1199] grpc_call_start_batch(call=0x20582c0, ops=0x2059370, nops=1, tag=0x204fcb0)
I0522 09:09:45.793334955   16441 call.c:1199] ops[0]: RECV_MESSAGE ptr=0x204f7b8
I0522 09:09:45.793385046   17073 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x204fcb0 OK
I0522 09:09:45.793895668   17072 completion_queue.c:223] RETURN_EVENT[0x1eb4470]: OP_COMPLETE: tag:0x1f11910 ERROR


  0 passing (919ms)
  1 failing

  1) Interop tests should pass server_streaming:
     Uncaught AssertionError: 3 === 4
      at ClientReadableStream.<anonymous> (/usr/local/google/home/ctiller/grpc/src/node/interop/interop_client.js:159:12)
      at ClientReadableStream.EventEmitter.emit (events.js:95:17)
      at /usr/local/google/home/ctiller/grpc/src/node/src/client.js:393:16
```

--- a/src/node/interop/interop_client.js
+++ b/src/node/interop/interop_client.js
@@ -154,12 +154,14 @@ function serverStreaming(client, done) {
                        arg.response_parameters[resp_index].size);
     resp_index += 1;
   });
-  call.on('status', function(status) {
-    assert.strictEqual(status.code, grpc.status.OK);
+  call.on('end', function() {
     assert.strictEqual(resp_index, 4);
     if (done) {
       done();
     }
+  });
+  call.on('status', function(status) {
+    assert.strictEqual(status.code, grpc.status.OK);
   });
 }
 

--- a/src/node/src/client.js
+++ b/src/node/src/client.js
@@ -80,11 +80,7 @@ function _write(chunk, encoding, callback) {
   var batch = {};
   batch[grpc.opType.SEND_MESSAGE] = this.serialize(chunk);
   this.call.startBatch(batch, function(err, event) {
-    if (err) {
-      callback(err);
-      return;
-    }
-    callback();
+    callback(err);
   });
 }
 

--- a/src/node/src/client.js
+++ b/src/node/src/client.js
@@ -80,7 +80,11 @@ function _write(chunk, encoding, callback) {
   var batch = {};
   batch[grpc.opType.SEND_MESSAGE] = this.serialize(chunk);
   this.call.startBatch(batch, function(err, event) {
-    callback(err);
+    if (err) {
+      // Something has gone wrong. Stop writing by failing to call callback
+      return;
+    }
+    callback();
   });
 }
 

--- a/src/node/src/client.js
+++ b/src/node/src/client.js
@@ -81,7 +81,7 @@ function _write(chunk, encoding, callback) {
   batch[grpc.opType.SEND_MESSAGE] = this.serialize(chunk);
   this.call.startBatch(batch, function(err, event) {
     if (err) {
-      // Something has gone wrong. Stop writing by failing to call callback
+      callback(err);
       return;
     }
     callback();
@@ -125,11 +125,9 @@ function _read(size) {
       self.finished = true;
       return;
     }
-    if (self.finished) {
-      self.push(null);
-      return;
-    }
     var data = event.read;
+    var deserialized = self.deserialize(data);
+    console.log(deserialized);
     if (self.push(self.deserialize(data)) && data !== null) {
       var read_batch = {};
       read_batch[grpc.opType.RECV_MESSAGE] = true;

--- a/src/node/src/client.js
+++ b/src/node/src/client.js
@@ -126,8 +126,6 @@ function _read(size) {
       return;
     }
     var data = event.read;
-    var deserialized = self.deserialize(data);
-    console.log(deserialized);
     if (self.push(self.deserialize(data)) && data !== null) {
       var read_batch = {};
       read_batch[grpc.opType.RECV_MESSAGE] = true;

